### DISCUSSION TITLE
контроллер игрока

### DISCUSSION THREAD

**Discussion by Pigzinzspace** [godot-engine-ru/shooter-3d#13] [Ideas]
**содержание документа** -  тут описаны и собраны параметры главного героя, анимаций главного героя  - что происходит и как это выглядит
 
**aнимации игрока от первого лица**
бег 
приседание
положение лежа
прыгает/спрыгивает 
падает сбитый с ног или откуда то например если его откинуло или бросил монстр 
поднимается с земли
оглушен/ослеплен
смена оружия, перезарядка оружия
движение руками при атаках ближнего боя,
движение руками при атаках дальнего боя 
взаимодействие открывает что-то (дверь с ноги или толкает плечом) или открывает за ручку, отвинчивает, использует ключ, перекусывает замок клещами, печатает на клавиатуре включает/выключает выключатель. 

**параметры игрока**
хиты
последняя выполненная миссия, или закрытый чекпоинт если они будут 
открытые секреты

**Comment by alex7850** [comment]
К топику о 
> Pigzinzspace: такой вопрос,  у нас будет режим ускорения(бега) и как он будет работать?
> варианты такие: 
> 0  - нет, не будет 
> 1 олдскульный  режим -  игрок двигается быстро, если нажать кнопку игрок двигается медленно, если еще раз нажать опять быстро.
> 2 режим с энергией - у игрока есть ресурс энергия, который постепенно восстанавливается  сам по себе, но расходуется при режиме бега.
> 3 уворот - режима бега нет, но есть увороты, которые позволяют быстро сместиться в сторону и избежать урон  или увернуться от контроля, они расходуют шкалу энергии квантами (например сразу 25, или 30, или 50%), а восстанавливается энергия постепенно.
> 4 комбинированный режим 2 и 3 - можно и ускоряться, и уворачиваться, быстрое движение уменьшает энергию постепенно, а уворот  отнимает фиксированные части.

### Кратко: начинать с неизменяемого "бега" в медленном-среднем темпе, но не исключать навешивания фич по мере разработки.

### Развернуто:
Считаю, нужно придерживаться принципов:
1) скорость перемещения должна быть достаточно медленной, чтобы были оправданы какие-то тактические маневры.
Например: В некоторых играх скорость игрока слишком высока и это часто делает бегание по периметру арены с распрыгой предпочтительнее интересному геймплею. В место этого игрок мог иметь время продумать ловушки для врагов и пути отступления

2) **easy to learn hard to master** - т.е. поощрять игрока за ловкое применение предметов, "трюков", использования левелдизайна какой-то прибавкой к скорости передвижения
Например: в некоторых играх есть неявные сразу для новичка паттерны передвижения в роде bunny hop, стрельба из оружия или подрыв бочки может ускорить отдачей, а грамотное подтягивание себя крюком дать взлететь.
[Пример скилового перемещения](https://www.reddit.com/r/godot/comments/sfb2ov/new_arena_mostly_done_for_my_duck_fps_quackout/)[
Пример 2](https://www.reddit.com/r/godot/comments/sezhp9/in_1_hour_were_hosting_a_play_testing_session_of/)
Естественно, не предлагаю превращать геймплей слишком в акробатический цирк

3) В геймдеве даже крупные студии с трудом могут планировать наперед механики. Любое внедрение фич требует геймдизайнерского тестирования и я не согласен с тем, чтобы прям железно их расписывать на год вперед. Не согласен, что мой подход приведет к потере времени, если будем изначально учитывать в коде возможные изменения.

**Reply by Pigzinzspace** [reply]
то есть уворот не нужен? 
в нашей игре не будет путей отступления,  у нас эта проблема решается тем что мобы спавнятся за спиной игрока, и это одна из причин почему они спавнятся, вторая причина что игрок бегает по комплексу туда сюда и мы не хотим  чтоб он заскучал по дороге назад.

**Reply by alex7850** [reply]
В самом начале разработки нет, но неплохо бы иметь. Другое дело, что его можно очень разными способами сделать, не только явным дешем на кнопку

**Reply by slapin** [reply]
уворот как QTE с вовремя нажатой клавшей если зомби схватил за задницу или набрасывается был бы неплох, в остальных случаях будет бесполезной функцией.

**Reply by slapin** [reply]
Вообще в AAA-проектах механики прототипируются и проходят много стадий прежде чем внедряются и кое-какие планы по механикам на ранней стадии все же делаются. С другой стороны ожидать у нас такого же я думаю бесполезно. Давайте делать все что можем, если механика кажется спорной - мы можем ее в меню включать-выключать и так решить проблемы с конфликтом механик :) так же это позволит проще прототипировать и даст свободу творчеству. Тем более чтобы понять подходит что-то или нет это сначала надо сделать, а выкидывать потом обидно и приводит к выгоранию...

**Reply by Pigzinzspace** [reply]
я тут выступаю с консервативных позиций,  я противник творчества и фантазий тут и того чтоб рассматривать этот проект как экспериментальный в плане идей и механик. Он не настолько интересен своим содержанием, чтоб тратить на него много времени и как то его варьировать. Это тестовый  проект, в плане на сколько мы способны совместно делать что-то быстро и четко. Мы берем простую тему полностью обрезаем все вариации и делаем максимально придерживаясь заданному, согласованному  шаблону.

**Reply by me2beats** [reply]
> в нашей игре не будет путей отступления, у нас эта проблема решается тем что мобы спавнятся за спиной игрока

Вот это не совсем понял, зачем спавнить мобов за спиной игрока.
Это же не хоррор; или есть шутаны где это норма?

**Comment by soundsbeard** [comment]
по поводу бега:
я считаю, что если мы делаем простой шутан, от которого игрок должен получать удовольствие, то он должен быть динамичным.
без бега динамично не получится. если я не прав, то зановите тайтлы, где без бега шутан получился таковым.

иметь постоянный бег без ограничений тож не комильфо, тк зомби просто не буду игрока догонять.
в общем я считаю, что нужен режим с энергией.
но это просто мнение как игрока, если решите иначе, то ок :+1:

**Reply by Pigzinzspace** [reply]
> зомби просто не буду игрока догонять.

зомби будут догонять,  потому что они бегут с постоянным разгоном как ракеты, рано или поздно догонят игрока, по карайней мере максимальная скорость зомби выше максимальной скорости игрока, просто убегать от зомби не получится, иначе смысл с ними драться вообще.

**Reply by soundsbeard** [reply]
в dying light **дефолтные** зомбаки вроде бы тоже игрока догнать не могут.
тем не менее, там всегда есть ситуации, когда надо именно прорываться через толпу зомби.
это просто как пример того, что даже неспособные догнать игрока зомби всё равно большая угроза.

**Reply by alex7850** [reply]
Есть целый жанр медленных шутеров без бега, где есть два режима "медленно" и "очень медленно". Вспомни кс 1.6, там было недостаточно экшена?
Конечно, обычно так делают в симуляторно тактических играх, но часто попадаются вполне попсовые реализации всяких resident evil, left4dead и даже в новых думах особо не поспринтуешь

Считаю, что дефолтная скорость должна чувствоваться достаточно быстрой, позволять уворачиваться от всяких проджектайлов, немного отступить от врага, но при этом вынуждать применять эффективно полный арсенал. Важно не приводить к ситуации, где бежать и отстреливаться будет ролять 90% случаев.

**Reply by slapin** [reply]
на самом деле если в беге ронять точность стрельбы на 70% уже будет давать выбор способа подхода к стрельбе и определенный тактический момент

**Reply by slapin** [reply]
> зомби будут догонять, потому что они бегут с постоянным разгоном как ракеты, рано или поздно догонят игрока, по карайней мере максимальная скорость зомби выше максимальной скорости игрока, просто убегать от зомби не получится, иначе смысл с ними драться вообще.

Убегание как тактический прием очень важно в хоррор-шутере, зомби должны быть страшными. Я бы даже добавил некоторые девайсы, облегчающие убегание, позволяющие зомби замедлить и укрытия, в которых зомби "не видят" если не видели куда спрятался...

**Reply by Pigzinzspace** [reply]
окей, как я понимаю никто особо не хочет уворотов, а по поводу спринтов мнения разделились

**Reply by Pigzinzspace** [reply]
> Убегание как тактический прием очень важно в хоррор-шутере, зомби должны быть страшными. 

ну у нас то позитивный шутер , у нас зомби это просто салат который мы крошим, мы очень нежны с психикой игрока никакого эмоционального хардкора - только позитив!

я бы сказал что эмоциональный рефернс нашей игры это "звездный десант"

**Reply by slapin** [reply]
По поводу скорости бега - это хороший кандидат на запихивание в меню как настройки, хотя бы по началу игры, потом сможем подобрать несколько интересных значений. Другой вопрост что если мы хотим вид от 3-го лица в какой-то перспективе, то нам будут нужны анимации и они будут определять скорость в первую очредь. Если же третьего лица ни в каком виде не будет, то скорость можно просто ограничить скоростью при которой капсула погружается наполовину в стену за физический тик чтобы сквозь стены было бегать нельзя. Остальное можно настраивать.

**Reply by Pigzinzspace** [reply]
это гуд идея чтоб не парится с дашем и спринтами, просто потестить с разной бефолтной скоростью

**Reply by Pigzinzspace** [reply]
мы хоти анимацию от первого лица от третьего нам критически усложнит анимацию, и ничего не даст

**Comment by slapin** [comment]
Динамичный шутан != непрерывный бег с прыжками по стенам, динамичность определяется в основном количеством событий в секунду и переменой напряжения / давления на игрока; бег это самый примитивный способ достижения этого, но например в Serious Sam и в Uncharted применяются другие способы нагнетания. И динамичным шутан делает перемена pacing'а - то есть после создания напряжения должно идти расслабление и это собственно и создает ощущение динамики, при линейном росте pacing'а многие неопытные игроки начинают "скучать". По сути если бег используется как способ регулировки pacing'а, это не очень правильная механика или не очень правильные ожидания от pacing'а (например игра от игрока ждет что он в этот момент будет искать цацки и лечиться, а игрок вместо этого рвется дальше или наоборот, от игрока ждут быстрый проскок по локации а он шкерится по укрытиям) такие моменты должны митигироваться или переключая интерес игрока или возможностью интересного прохода локации несколькими способами с той скоростью, с какой у игрока есть настроение.

**Comment by Einshtein12** [comment]
Hello, это я @Эйнштейн#0730, предлагаю создать file settings где будут храниться характеристики игрока которые можно прокачивать(высота прыжка, урон, hp и т.д.)

**Reply by Pigzinzspace** [reply]
у нас же вроде шутер а не рпг

**Reply by Einshtein12** [reply]
А в чем отличие? Пж без fp

**Reply by Pigzinzspace** [reply]
в рпг можно прокачивать в шутере только убивать, в рпг ты убиваешь из выгоды а в шутере только  из удовольствия.

**Comment by slapin** [comment]
Нам все равно базовые параметры понадобятся. Это ж все настраивать придется.

**Reply by Pigzinzspace** [reply]
более развернуто можешь написать ?

**Reply by slapin** [reply]
Ну например базовый урон игрока, базовое здоровье и тп. Есть набор геймплейных констант которые придется настраивать.
Ну и коэффициенты для каждого баффа, брони и тп. Это должно жить в одном месте для удобства тюнинга UX'а.

**Reply by Pigzinzspace** [reply]
не вопрос, но это уже не к игре, в игре это все будет константами

**Reply by slapin** [reply]
Это к проекту, нужно обеспечить настройку.
